packages/test/stochastic-test-utils/src/generators.ts

```diff
@@ -0,0 +1,355 @@
+            ? (tOrGenerator as Generator<T, TState>)(state)
+            : (tOrGenerator as unknown as T);
+    };
+}
+
+/**
+ * Higher-order generator operator which creates a new generator producing the first `n` elements of `generator`.
+ */
+export function take<T, TState>(n: number, generator: Generator<T, TState>): Generator<T, TState> {
+    let count = 0;
+    return (state) => {
+        if (count < n) {
+            count++;
+            return generator(state);
+        }
+        return done;
+    };
+}
+
+/**
+ * @returns a deterministic generator that always returns the items of `contents` in order.
+ */
+export function generatorFromArray<T, TAdditionalState>(contents: T[]): Generator<T, TAdditionalState> {
+    let index = -1;
+    return () => {
+        if (index < contents.length) {
+            index++;
+            return contents[index] ?? done;
+        }
+        return done;
+    };
+}
+
+/**
+ * Higher-order generator operator which exhausts each input generator sequentially before moving on to the next.
+ */
+export function chain<T, TState>(...generators: Generator<T, TState>[]): Generator<T, TState> {
+    let currentIndex = 0;
+    return (state) => {
+        while (currentIndex < generators.length) {
+            const generator = generators[currentIndex];
+            const result = generator(state);
+            if (result !== done) {
+                return result;
+            } else {
+                currentIndex++;
+            }
+        }
+        return done;
+    };
+}
+
+/**
+ * Higher-order generator operator which exhausts each input generator sequentially before moving on to the next.
+ */
+export function chainIterables<T, TState>(generators: Generator<Generator<T, TState>, void>): Generator<T, TState> {
+    let currentGenerator = generators();
+    return (state) => {
+        while (currentGenerator !== done) {
+            const result = currentGenerator(state);
+            if (result !== done) {
+                return result;
+            }
+
+            currentGenerator = generators();
+        }
+
+        return done;
+    };
+}
+
+/**
+ * Interleaves outputs from `generator1` and `generator2`.
+ * By default outputs are taken one at a time, but can be controlled with `numOps1` and `numOps2`.
+ * This is useful in stochastic tests for producing a certain operation (e.g. "validate" or "synchronize") at a
+ * defined interval.
+ *
+ * Exhausts both input generators before terminating.
+ * @example
+ * ```typescript
+ * // Assume gen1 produces 1, 2, 3, ... and gen2 produces "a", "b", "c", ...
+ * interleave(gen1, gen2) // 1, a, 2, b, 3, c, ...
+ * interleave(gen1, gen2, 2) // 1, 2, a, 3, 4, b, 5, 6, c, ...
+ * interleave(gen1, gen2, 2, 3) // 1, 2, a, b, c, 3, 4, d, e, f, ...
+ * ```
+ */
+export function interleave<T, TState>(
+    generator1: Generator<T, TState>,
+    generator2: Generator<T, TState>,
+    numOps1 = 1,
+    numOps2 = 1,
+): Generator<T, TState> {
```

Comment:
ah, annoying. Git marked this file as a delete rather than a move. Probably partially due to formatting changes. There's a way to be explicit about this IIRC, but hopefully it's not too bad to review.

This function and the async variety actually do have a bug fix applied that I found while unit testing / explained in the comment below. Other than that and an added specification in the doc comment (L125 here) this file should be unchanged semantically.